### ISSUE TITLE
CUDA error: CUBLAS_STATUS_INVALID_VALUE when calling `cublasSgemm

### ISSUE THREAD

**Issue by velastin** [ultralytics/ultralytics#693]
I get the error when trying to run a prediction, just after installation (in fact, the error occurs when trying to do anything with the CLI or with PIP). Installation (perhaps the problem is conda? version is conda 22.9.0, running in Linux):
conda update -n base -c defaults conda
conda create --no-default-packages -n sergio_yolov8 python=3.8 ipython
conda activate sergio_yolov8
pip install ultralytics
yolo predict model=yolov8n.pt imgsz=640 conf=0.25

Output
Downloading https://github.com/ultralytics/assets/releases/download/v0.0.0/yolov8n.pt to yolov8n.pt...
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████| 6.23M/6.23M [00:00<00:00, 11.7MB/s]

WARNING ⚠️ 'source=' is missing. Using default 'source=https://ultralytics.com/images/bus.jpg'.
Ultralytics YOLOv8.0.22 🚀
Traceback (most recent call last):
  File "/home/sergio/anaconda3/envs/sergio_yolov8/bin/yolo", line 8, in <module>
    sys.exit(entrypoint())
  File "/home/sergio/anaconda3/envs/sergio_yolov8/lib/python3.8/site-packages/ultralytics/yolo/cfg/__init__.py", line 265, in entrypoint
    getattr(model, mode)(**overrides)
  File "/home/sergio/anaconda3/envs/sergio_yolov8/lib/python3.8/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "/home/sergio/anaconda3/envs/sergio_yolov8/lib/python3.8/site-packages/ultralytics/yolo/engine/model.py", line 142, in predict
    self.predictor.setup_model(model=self.model)
  File "/home/sergio/anaconda3/envs/sergio_yolov8/lib/python3.8/site-packages/ultralytics/yolo/engine/predictor.py", line 236, in setup_model
    self.model = AutoBackend(model, device=device, dnn=self.args.dnn, data=self.args.data, fp16=self.args.half)
  File "/home/sergio/anaconda3/envs/sergio_yolov8/lib/python3.8/site-packages/ultralytics/nn/autobackend.py", line 66, in __init__
    model = model.fuse() if fuse else model
  File "/home/sergio/anaconda3/envs/sergio_yolov8/lib/python3.8/site-packages/ultralytics/nn/tasks.py", line 100, in fuse
    m.conv = fuse_conv_and_bn(m.conv, m.bn)  # update conv
  File "/home/sergio/anaconda3/envs/sergio_yolov8/lib/python3.8/site-packages/ultralytics/yolo/utils/torch_utils.py", line 133, in fuse_conv_and_bn
    fusedconv.bias.copy_(torch.mm(w_bn, b_conv.reshape(-1, 1)).reshape(-1) + b_bn)
RuntimeError: CUDA error: CUBLAS_STATUS_INVALID_VALUE when calling `cublasSgemm( handle, opa, opb, m, n, k, &alpha, a, lda, b, ldb, &beta, c, ldc)`
Sentry is attempting to send 2 pending error messages
Waiting up to 2 seconds
Press Ctrl-C to quit

**Comment by SV1367** [comment]
The issue is of course avoided by running in a docker

**Comment by lukazso** [comment]
In my case the issue was that I somehow installed a torch for cuda 11.7, however I am still running cuda 11.5 on my notebook. You might wanna look for a mismatch like this.

**Comment by pderrenger** [comment]
@lukazso it appears that the issue might be related to a mismatch between the installed version of PyTorch and the running version of CUDA. When encountering the "CUDA error: CUBLAS_STATUS_INVALID_VALUE" when calling `cublasSgemm` error, it's essential to ensure compatibility between the installed PyTorch version and the CUDA version on the system. Checking for mismatches in versions, such as installing a PyTorch version for a different CUDA version, could help resolve the issue. Additionally, running the environment in a Docker container can effectively bypass the compatibility concerns.

If you encounter any further issues or require additional assistance, feel free to reach out!

**Comment by zjykzj** [comment]
> fusedconv.bias.copy_(torch.mm(w_bn, b_conv.reshape(-1, 1)).reshape(-1) + b_bn)

@velastin @pderrenger @SV1367 @lukazso I encountered the same error, but in other code.

```shell
...
...
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically...
optimizer: AdamW(lr=0.001667, momentum=0.9) with parameter groups 66 weight(decay=0.0), 77 weight(decay=0.0005), 76 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 8 dataloader workers
Logging results to /data/zj/renjie_train/rj_train/tmp/1944954995169890305_train_1752549276/model_train
Starting training for 100 epochs...

      Epoch    GPU_mem   box_loss   seg_loss   cls_loss   dfl_loss  Instances       Size
      1/100       4.5G     0.6546       1.88      2.243      1.227         36        640: 100%|██████████| 17/17 [00:08<00:00,  1.92it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Mask(P          R      mAP50  mAP50-95):   0%|          | 0/3 [00:00<?, ?it/s]
Traceback (most recent call last):
  File "segment_train.py", line 100, in <module>
    main()
  File "segment_train.py", line 96, in main
    trainer.train()
  File "/data/zj/renjie_train/rj_train/framework/ultralytics/ultralytics/engine/trainer.py", line 214, in train
    self._do_train(world_size)
  File "/data/zj/renjie_train/rj_train/framework/ultralytics/ultralytics/engine/trainer.py", line 441, in _do_train
    self.metrics, self.fitness = self.validate()
  File "/data/zj/renjie_train/rj_train/framework/ultralytics/ultralytics/engine/trainer.py", line 638, in validate
    metrics = self.validator(self)
  File "/data/miniconda3/envs/py38/lib/python3.8/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "/data/zj/renjie_train/rj_train/framework/ultralytics/ultralytics/engine/validator.py", line 221, in __call__
    self.loss += model.loss(batch, preds)[1]
  File "/data/zj/renjie_train/rj_train/framework/ultralytics/ultralytics/nn/tasks.py", line 299, in loss
    return self.criterion(preds, batch)
  File "/data/zj/renjie_train/rj_train/framework/ultralytics/ultralytics/utils/loss.py", line 301, in __call__
    pred_bboxes = self.bbox_decode(anchor_points, pred_distri)  # xyxy, (b, h*w, 4)
  File "/data/zj/renjie_train/rj_train/framework/ultralytics/ultralytics/utils/loss.py", line 196, in bbox_decode
    pred_dist = pred_dist.view(b, a, 4, c // 4).softmax(3).matmul(self.proj.type(pred_dist.dtype))
RuntimeError: CUDA error: CUBLAS_STATUS_INVALID_VALUE when calling `cublasSgemv(handle, op, m, n, &alpha, a, lda, x, incx, &beta, y, incy)`
```

I have tried many ways to solve this problem,

1. Verify if it is an error in the dataset
2. Verify if it is an error in the input matrix during matrix multiplication calculation
3. Adjust CUDNN (Or, to put it another way (including `pytorch/cuda/cudnn`), mentioned by others)
4. Modify the code (**this step solves the problem**)

## Verify if it is an error in the dataset

```python3
# check_yolo_dataset.py
import os

# 设置路径
label_train_dir = 'labels/train'
label_val_dir = 'labels/val'

# 类别总数（根据你的 dataset 修改）
num_classes = 2

def check_label_file(file_path):
    with open(file_path, 'r') as f:
        lines = f.readlines()

    if not lines:
        print(f"[ERROR] Empty label file: {file_path}")
        return False

    for line in lines:
        parts = line.strip().split()
        if len(parts) < 5:
            print(f"[ERROR] Too few values in line '{line}' in file: {file_path}")
            return False

        try:
            class_id = int(parts[0])
            coords = list(map(float, parts[1:5]))
        except ValueError as e:
            print(f"[ERROR] Invalid value format in line '{line}' in file: {file_path} - {e}")
            return False

        if class_id < 0 or class_id >= num_classes:
            print(f"[ERROR] Invalid class_id {class_id} in file: {file_path}")
            return False

        for coord in coords:
            if coord < 0 or coord > 1:
                print(f"[ERROR] Coordinate out of range [0,1]: {coord} in file: {file_path}")
                return False

    return True


def scan_label_folder(label_dir):
    print(f"Scanning folder: {label_dir}")
    count = 0
    error_count = 0

    for filename in os.listdir(label_dir):
        if filename.endswith('.txt'):
            file_path = os.path.join(label_dir, filename)
            count += 1
            if not check_label_file(file_path):
                error_count += 1

    print(f"Checked {count} files in {label_dir}, found {error_count} errors.\n")
    return error_count == 0


if __name__ == '__main__':
    print("🔍 Checking YOLO Dataset for anomalies...\n")

    ok1 = scan_label_folder(label_train_dir)
    ok2 = scan_label_folder(label_val_dir)

    if ok1 and ok2:
        print("✅ All label files are valid.")
    else:
        print("❌ Some label files have issues. Please fix them.")
```

## Verify if it is an error in the input matrix during matrix multiplication calculation

```python3
    def bbox_decode(self, anchor_points, pred_dist):
        """Decode predicted object bounding box coordinates from anchor points and distribution."""
        if self.use_dfl:
            b, a, c = pred_dist.shape  # batch, anchors, channels

            print("pred_dist.device:", pred_dist.device)
            print("self.proj.device:", self.proj.device)
            print("self.proj.isnan().any():", self.proj.isnan().any())
            print("self.proj.isinf().any():", self.proj.isinf().any())
            
            pred_dist = pred_dist.view(b, a, 4, c // 4).softmax(3).matmul(self.proj.type(pred_dist.dtype))
            # pred_dist = pred_dist.view(b, a, c // 4, 4).transpose(2,3).softmax(3).matmul(self.proj.type(pred_dist.dtype))
            # pred_dist = (pred_dist.view(b, a, c // 4, 4).softmax(2) * self.proj.type(pred_dist.dtype).view(1, 1, -1, 1)).sum(2)
        return dist2bbox(pred_dist, anchor_points, xywh=False)
```

## Adjust CUDNN

Actually, I have tried training in both Docker and host environments, but for this specific dataset and training configuration, there will always be training errors. One of the environment configurations is as follows:

```shell
$ python3 -m torch.utils.collect_env
Collecting environment information...
PyTorch version: 1.12.1
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.10.2
Libc version: glibc-2.27

Python version: 3.8.20 (default, Oct  3 2024, 15:24:27)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.4.0-150-generic-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 11.8.89
GPU models and configuration:
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 3090

Nvidia driver version: 545.29.02
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.9.7
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.9.7
/usr/local/cuda-11.8/targets/x86_64-linux/lib/libcudnn.so.8.9.7
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] torch==1.12.1
[pip3] torchaudio==0.12.1
[pip3] torchvision==0.13.1
[conda] blas                      1.0                         mkl    defaults
[conda] cudatoolkit               11.3.1               h2bc3f7f_2    defaults
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2023.1.0         h213fc3f_46344    defaults
[conda] mkl-service               2.4.0            py38h5eee18b_1    defaults
[conda] mkl_fft                   1.3.8            py38h5eee18b_0    defaults
[conda] mkl_random                1.2.4            py38hdb19cb5_0    defaults
[conda] numpy                     1.24.3           py38hf6e8229_1    defaults
[conda] numpy-base                1.24.3           py38h060ed82_1    defaults
[conda] pytorch                   1.12.1          py3.8_cuda11.3_cudnn8.3.2_0    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                0.12.1               py38_cu113    pytorch
[conda] torchvision               0.13.1               py38_cu113    pytorch
```

I also tried CPU training and excluding the use of CUDNN, both of which were executed normally

```python3
import torch
torch.backends.cudnn.enabled = False
print("Disabled cuDNN")
```

## Modify the code (**this step solves the problem**)

After conducting the above research, I began to break down the implementation of specific error codes step by step and found that it was caused by the shape of the input matrix not meeting the requirements in matrix multiplication operations

```python3
pred_dist = pred_dist.view(b, a, 4, c // 4).softmax(3).matmul(self.proj.type(pred_dist.dtype))
```

Modify to the following implementation, forcing matrix transformation operation

```python3
pred_dist = pred_dist.view(b, a, 4, c // 4).softmax(3)  # shape: (b, a, 4, 16)
proj = self.proj.type(pred_dist.dtype).reshape(16, 1)   # shape: (16, 1)
pred_dist = pred_dist.matmul(proj).squeeze(-1)          # shape: (b, a, 4)
```

Yes, this operation solved the problem !!!

 In fact, this training error was a very occasional behavior. I have conducted previous training normally, and for this training, setting up small batch (like `--batch 1/2`) training can be carried out normally. 

In summary, it should be noted that when implementing the broadcast mechanism at the bottom of PyTorch/CUDA, there may be different implementations for data deformations of different sizes, resulting in occasional error occurrences. @glenn-jocher  

* #10409
* [Error when running inference on V100 GPU](https://github.com/ultralytics/yolov5/issues/10046)
* [RuntimeError: CUDA error: CUBLAS_STATUS_INVALID_VALUE when calling cublasSgemm( handle, opa, opb, m, n, k, &alpha, a, lda, b, ldb, &beta, c, ldc) #9986](https://github.com/ultralytics/yolov5/issues/9986)

**Comment by glenn-jocher** [comment]
@zjykzj thank you for the comprehensive investigation and solution! This CUDA error is indeed related to PyTorch/CUDA version compatibility issues as you've discovered.

The proper solution is to ensure your PyTorch and CUDA versions are compatible rather than modifying the core code. You can verify compatibility by running `import torch; print(torch.cuda.is_available())` and checking your versions match.

For immediate resolution, install the correct PyTorch version for your CUDA setup:
```bash
conda install pytorch torchvision torchaudio pytorch-cuda=11.8 -c pytorch -c nvidia
```

Replace `11.8` with your actual CUDA version. This approach maintains code integrity while resolving the underlying compatibility issue that causes the matrix multiplication errors you observed.

**Comment by zjykzj** [comment]
> [@zjykzj](https://github.com/zjykzj) thank you for the comprehensive investigation and solution! This CUDA error is indeed related to PyTorch/CUDA version compatibility issues as you've discovered.
> 
> The proper solution is to ensure your PyTorch and CUDA versions are compatible rather than modifying the core code. You can verify compatibility by running `import torch; print(torch.cuda.is_available())` and checking your versions match.
> 
> For immediate resolution, install the correct PyTorch version for your CUDA setup:
> 
> conda install pytorch torchvision torchaudio pytorch-cuda=11.8 -c pytorch -c nvidia
> Replace `11.8` with your actual CUDA version. This approach maintains code integrity while resolving the underlying compatibility issue that causes the matrix multiplication errors you observed.

Hi @glenn-jocher, the current training environment should meet the training requirements of ultralytics. The previous training was conducted normally, but this time there was a sudden training anomaly.

```shell
(py38) user@user-X11DAi-N:/data/zj/renjie_train$ python3
Python 3.8.20 (default, Oct  3 2024, 15:24:27)
[GCC 11.2.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> torch.__version__
'1.12.1'
>>> torch.cuda.is_available()
True
>>> import torchvision
>>> torchvision.__version__
'0.13.1'
>>>
(py38) user@user-X11DAi-N:/data/zj/renjie_train$ yolo checks
Ultralytics 8.3.100 🚀 Python-3.8.20 torch-1.12.1 CUDA:0 (NVIDIA GeForce RTX 3090, 24260MiB)
Setup complete ✅ (48 CPUs, 125.6 GB RAM, 862.2/937.8 GB disk)

OS                  Linux-5.4.0-150-generic-x86_64-with-glibc2.17
Environment         Linux
Python              3.8.20
Install             git
Path                /data/miniconda3/envs/py38/lib/python3.8/site-packages/ultralytics
RAM                 125.56 GB
Disk                862.2/937.8 GB
CPU                 Intel Xeon Silver 4214 2.20GHz
CPU count           48
GPU                 NVIDIA GeForce RTX 3090, 24260MiB
GPU count           2
CUDA                11.3

numpy               ✅ 1.24.3<=2.1.1,>=1.23.0
matplotlib          ✅ 3.7.5>=3.3.0
opencv-python       ❌ (not installed)>=4.6.0
pillow              ✅ 10.4.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.10.0>=1.4.1
torch               ✅ 1.12.1>=1.8.0
torch               ✅ 1.12.1!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.13.1>=0.9.0
tqdm                ✅ 4.65.2>=4.64.0
psutil              ✅ 6.1.1
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.0.3>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.13>=2.0.0
```

for torch/torchvision install, I choose to download the whl then install locally

```text
torch-1.12.1+cu113-cp38-cp38-linux_x86_64.whl
torchvision-0.13.1+cu113-cp38-cp38-linux_x86_64.whl
```

**Comment by zjykzj** [comment]
> [@zjykzj](https://github.com/zjykzj) thank you for the comprehensive investigation and solution! This CUDA error is indeed related to PyTorch/CUDA version compatibility issues as you've discovered.
> 
> The proper solution is to ensure your PyTorch and CUDA versions are compatible rather than modifying the core code. You can verify compatibility by running `import torch; print(torch.cuda.is_available())` and checking your versions match.
> 
> For immediate resolution, install the correct PyTorch version for your CUDA setup:
> 
> conda install pytorch torchvision torchaudio pytorch-cuda=11.8 -c pytorch -c nvidia
> Replace `11.8` with your actual CUDA version. This approach maintains code integrity while resolving the underlying compatibility issue that causes the matrix multiplication errors you observed.

For the modification of `bbox_decode()` in the `v8DetectionLoss`, my final implementation method is as follows

```
    def bbox_decode(self, anchor_points, pred_dist):
        """Decode predicted object bounding box coordinates from anchor points and distribution."""
        if self.use_dfl:
            b, a, c = pred_dist.shape  # batch, anchors, channels

            pred_dist = pred_dist.view(b, a, 4, c // 4).softmax(3).matmul(
                self.proj.reshape(c // 4, 1).type(pred_dist.dtype)).squeeze(-1)
            # pred_dist = pred_dist.view(b, a, 4, c // 4).softmax(3).matmul(self.proj.type(pred_dist.dtype))
        return dist2bbox(pred_dist, anchor_points, xywh=False)
```

**Comment by glenn-jocher** [comment]
@zjykzj thanks for the detailed analysis and the working solution! Your investigation shows this is indeed a rare edge case where specific batch sizes and tensor shapes can trigger CUDA matrix multiplication errors even with compatible PyTorch/CUDA versions.

Your modification to explicitly reshape the projection matrix before multiplication is a reasonable workaround. The `self.proj.reshape(c // 4, 1)` approach ensures the matrix dimensions are explicit for the matmul operation, which avoids the broadcast mechanism issues you identified.

Since this affects training stability for certain configurations, we'll consider incorporating a similar fix into the main codebase. The solution maintains the same mathematical operation while being more explicit about tensor shapes, which should resolve the intermittent CUDA errors without impacting performance.

For now, your local modification should allow you to continue training normally. Thanks for the thorough debugging - this type of investigation really helps improve the robustness of the codebase!
